sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2StreamingScanSupportCheck.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.streaming.{StreamingRelation, StreamingRelationV2}
+import org.apache.spark.sql.sources.v2.TableCapability.{CONTINUOUS_READ, MICRO_BATCH_READ}
+
+/**
+ * This rules adds some basic table capability check for streaming scan, without knowing the actual
+ * streaming execution mode.
+ */
+object V2StreamingScanSupportCheck extends (LogicalPlan => Unit) {
+  import DataSourceV2Implicits._
+
+  override def apply(plan: LogicalPlan): Unit = {
```

Comment:
I think these checks should also include a validation for individual tables. If a table doesn't support streaming at all, a more helpful error message is that a specific table doesn't support streaming, not just that there isn't a streaming mode that can handle all of the sources.

Reply:
If we add this check here, it will never be hit because we already checked it earlier in `DataStreamReader`. Like I explained, it's non-trivial to move the check from `DataStreamReader`  to here, because it's coupled with the v2 -> v1 fallback logic.

Reply:
I think eventually we will have an analyzer rule that checks streaming scan capability and fallback to v1 if necessary. This checker rule is not suitable because it just traverses the plan, not returning a new plan. So we can't implement the fallback logic in this rule.

Reply:
Even if it will never be hit, it should still be in the analyzer so that plans that are created through other paths in the future are still checked. The idea is to avoid relying on a particular way of creating a logical plan to validate logical plans.

Reply:
that's why I opened https://issues.apache.org/jira/browse/SPARK-27483

plans that are created through other APIs(not `DataStreamReader`) still need the fallback logic, which can not be done within this checker rule.

Reply:
Why would this check depend on fallback logic? These checks run after resolution rules have reached a fixed point. If there is a streaming DSv2 relation in the plan, fallback should already be done. Fallback logic is separate and this check can be done here.

Reply:
> Why would this check depend on fallback logic?

It's the opposite. The fallback logic depends on the check. That said, the future analyzer rule would do 3 things:
1. find the v2 streaming relation in the plan, check scan capability
2. if check failed, fallback to v1 relation, and use the v1 relation to replace the v2 streaming relation.
3. if fallback is not applicable, fail

You can see that, it's not a simple check anymore, which can not be done within this simpler checker rule (`LogicalPlan => Unit`).

Reply:
@cloud-fan, let's get the validation in now.

I don't think that the fallback rule should be implemented as you describe. I think it should be done in 2 parts. The rule to fallback and update the plan, and a validation that all sources support streaming. Spark should not combine transform rules and validations. There are a couple of reasons for this principle:

1. Validations are used to ensure that the query is valid *and* to ensure that rules are run correctly. If the transform rule is added to the analyzer in a single-run batch, we want validation to catch that. These checks catch errors in Spark, too.
2. Rules should be as small as possible and focused on a single task. The fallback rule should not fail analysis if it doesn't know what to do because some other rule may be added later that does. For example, what if we build an adapter from continuous execution to micro-batch execution for a source?

So we will need a validation rule either way. When the fallback rule runs and can't fix the problem, this check should be what fails the plan.

Reply:
added.